### PR TITLE
*: Split etcd into 3 modules: server, client (c), integration

### PR DESCRIPTION
**Pull request by ptabor**
*: Split etcd into 3 modules:
  -   go.etcd.io/etcd/v3 - server
  -   go.etcd.io/etcd/c/v3 - client code
  -   go.etcd.io/etcd/integration/v3 - integration tests

This was discussed an community meeting on 2020/07/30. The goal is to
minimize number of dependencies that our network customers needs to
carry and clearly define packets that needs to stay backward compatible.

Main files moves into ./c
  Clients:
    ./{client,clientv3}
  API protos:
    ./etcdserver/api/membership/membershippb
    ./etcdserver/api/v3rpc/rpctypes
    ./auth/authpb
    ./mvcc/mvccpb
    ./etcdserver/etcdserverpb
  Integration tests:
    ./client{v3}/{integration,concurrency} -> ./integration/client{v3}/{integration,concurrency}

Tested using:

```
  PASSES="unit integration functional e2e" time ./test -v

  ./build

  (go build ./... && go test ./...)
  (cd c; go build ./... && go test ./...)
  (cd integration; go build ./... && go test ./...)

  ./scripts/genproto.sh

  make

  go mod tidy
  go mod vendor
```

Futher steps to consider:
  - migrate etcdctl into (c) - client code. This requires spliting some   admin funtionalities that operate directly on etcd server files.
  - move tests/e2e and functional into integration dir 

Please see/comment the attached [doc](https://docs.google.com/document/d/19UvKD7by_fEkzLMRi-QNSKEed4kYAW286S3DsNRPDOM/edit)

### CODE REVIEW

**Comment by ptabor** [comment]
\cc @gyuho  
\cc @jpbetz 

This is the PR that is separating the client code from the server code - discussed on the last community meeting. Please review.

**Comment by ptabor** [comment]
Travis is flaky on --- FAIL: TestGetTokenWithoutAuth (39.16s) - that is addressed in https://github.com/etcd-io/etcd/pull/12200

\cc @wojtek-t  FYI

**Comment by ptabor** [comment]
More about motivation and considerations in [modularization of etcd doc](https://docs.google.com/document/d/19UvKD7by_fEkzLMRi-QNSKEed4kYAW286S3DsNRPDOM/edit)

**Comment by gyuho** [comment]
@ptabor Can we have separate commits? It's hard to review :)

**Comment by ptabor** [comment]
@gyuho I will split when I'm back from vacations in a week. 
Majority of changes here are file moves and import changes. 

Interesting changes. go to:
  - go.mod files (./go.mod  ./c/go.mod. ./integration/go.mod)
  - ./test file

I would appreciate if could take a look at: ./test file and the overall directory layout structure in the meantime.

**Comment by ptabor** [comment]
@gyuho I splitted the change as requested. PTAL

**Comment by ptabor** [comment]
@gyuho @jpbetz @jingyih @wenjiaswe - please review

File: c/etcdserver/etcdserverpb/raft_internal.pb.go [hunk 1]
```diff
@@ -2270,69 +2270,68 @@ var (
 func init() { proto.RegisterFile("raft_internal.proto", fileDescriptorRaftInternal) }
 
 var fileDescriptorRaftInternal = []byte{
```
Comment (jpbetz) [inline c/etcdserver/etcdserverpb/raft_internal.pb.go:2272]: Why did this change? Same question for the other `.pb.go` files that changed.

File: scripts/genproto.sh [hunk 2]
```diff
@@ -10,13 +10,13 @@ if ! [[ "$0" =~ scripts/genproto.sh ]]; then
 	exit 255
 fi
 
-if [[ $(protoc --version | cut -f2 -d' ') != "3.7.1" ]]; then
-	echo "could not find protoc 3.7.1, is it installed + in PATH?"
-	exit 255
-fi
+#if [[ $(protoc --version | cut -f2 -d' ') != "3.7.1" ]]; then
```
Comment (jpbetz) [inline scripts/genproto.sh:13]: Re-enable or add comment explaining why we don't need this any more?

File: test [hunk 3]
```diff
@@ -50,8 +50,8 @@ USERPKG=${PKG:-}
 COVER=${COVER:-"-cover"}
 
 # Hack: gofmt ./ will recursively check the .git directory. So use *.go for gofmt.
-IGNORE_PKGS="(vendor/|etcdserverpb|rafttest|gopath.proto|v3lockpb|v3electionpb)"
-INTEGRATION_PKGS="(integration|tests/e2e|contrib|functional)"
+IGNORE_PKGS="(^|/)(vendor|etcdserverpb|rafttest|gopath.proto|v3lockpb|v3electionpb|c)(/|$)"
```
Comment (jpbetz) [inline test:53]: I don't think really want to exclude the entire `c` module from gofmt.

File: test [hunk 4]
```diff
@@ -50,8 +50,8 @@ USERPKG=${PKG:-}
...
+INTEGRATION_PKGS="(^|/)(integration|tests/e2e|contrib|functional|[.])(/|$)"
```
Comment (jpbetz) [inline test:54]: Thanks for fixing the regex matching to test full path parts only.

File: test [hunk 5]
```diff
@@ -172,27 +180,34 @@ function integration_pass {
...
 		integration_extra "$@"
 	else
 		if [ -z "${USERPKG}" ]; then
-			INTEGTESTPKG=("${REPO_PATH}/integration"
-						  "${REPO_PATH}/client/integration"
-						  "${REPO_PATH}/clientv3/integration"
-						  "${REPO_PATH}/contrib/raftexample")
+			INTEGTESTPKG=("${REPO_PATH}/contrib/raftexample")
+			go test -mod=mod -timeout "${USERTIMEOUT}" -v -cpu "${TEST_CPUS}" "${RUN_ARG}"  "$@" "${INTEGTESTPKG[@]}"
+            		(
```
Comment (jpbetz) [inline test:192]: nit: indentation of this line is odd, doesn't match 195

File: etcdserver/api/v3lock/v3lockpb/v3lock.pb.go [hunk 6]
```diff
@@ -934,26 +934,26 @@ var (
 func init() { proto.RegisterFile("v3lock.proto", fileDescriptorV3Lock) }
 
 var fileDescriptorV3Lock = []byte{
```
Comment (jpbetz) [inline etcdserver/api/v3lock/v3lockpb/v3lock.pb.go:936]: Here also: why the protobuf changes?

File: vendor/modules.txt [hunk 7]
```diff
@@ -27,7 +27,7 @@ github.com/dustin/go-humanize
 github.com/etcd-io/gofail/runtime
 # github.com/fatih/color v1.7.0
 ## explicit
-# github.com/gogo/protobuf v1.2.1
+# github.com/gogo/protobuf v1.2.1 => github.com/gogo/protobuf v1.2.1
```
Comment (jpbetz) [inline vendor/modules.txt:30]: Seems unnecessary.

File: vendor/modules.txt [hunk 8]
```diff
@@ -228,3 +256,6 @@ gopkg.in/yaml.v2
 # sigs.k8s.io/yaml v1.1.0
 ## explicit
 sigs.k8s.io/yaml
+# github.com/gogo/protobuf => github.com/gogo/protobuf v1.2.1
+# go.etcd.io/etcd/c/v3 => ./c
+# go.etcd.io/etcd/integration/v3 => ./SHOULD_NOT_BE_DEPENDENCY
```
Comment (jpbetz) [inline vendor/modules.txt:261]: Interesting. How does this line get added to modules.txt? Is it added manually or do we have some way to make sure it gets added automatically?

File: c/etcdserver/etcdserverpb/raft_internal.pb.go (hunk 1 above)
Comment (ptabor) [reply c/etcdserver/etcdserverpb/raft_internal.pb.go:2272]: It's just the result of running protoc on the proto file in a new directory. 

File location is stored as part of [FileDescriptorProto](https://github.com/protocolbuffers/protobuf/blob/f5bdd7cd56aa86612e166706ed8ef139db06edf2/src/google/protobuf/descriptor.proto#L63) and after gziping it leads to a different payload.

File: test (hunk 4 above)
Comment (ptabor) [reply test:54]: Your welcome

File: scripts/genproto.sh (hunk 2 above)
Comment (ptabor) [reply scripts/genproto.sh:13]: Re-enabled. Good catch.

Would be happy to add a comment why we need this. 
Lookled over changes 3.7..3.12, and there seems nothing go related:  https://github.com/protocolbuffers/protobuf/releases?before=v3.7.1. 

I assume its to guarantee better repeatability for patch releases.

File: c/.gitignore [hunk 9]
```diff
@@ -0,0 +1,2 @@
+# git mod vendor will regenerate the directory
+vendor/
```
[the remark this answers is not in the document]
Comment (ptabor) [reply c/.gitignore:2]: I moved the cleanup of vendor directory to a separate PR:  https://github.com/etcd-io/etcd/pull/12279 
I performed there multiple tests and: 
   - **neither "go mod fetch"** nor "go mod vendor" **is needed**.   "go build ..."   or "./build" works of fresh clients (no $GOPATH)
   - Performance is highly improved. Even if we wipe $GOPATH/pkg,  the **"./build" takes <10s without vendor and 
     ~34s with vendor**. I think that go mod must download pre-built or pre-analyzed artifacts to have such great improvement. 
   - Multiple modules within etcd or in different repositories would share the same pre-loaded modules. 

Please let me know in the PR whether I'm missed any benefits / scenarios to keep the ./vendor dir.

[the remark this answers is not in the document]
Comment (jpbetz) [reply c/.gitignore:2]: This is very promising. By removing vendor we are adding a build dependency on proxy.golang.org (and requiring a network connection to pull from it). This seems acceptable to me given that the proxy is quite mature at this point. I've added a comment about this on #12279 and nudged a couple maintainers. We can continue the conversation over there.

File: vendor/modules.txt (hunk 8 above)
Comment (ptabor) [reply vendor/modules.txt:261]: It's a consequence of: https://github.com/etcd-io/etcd/pull/12204/commits/053201df78b6dc034ca20101569c900d2948ebe9#diff-37aff102a57d3d7b797f152915a6dc16R54 in go.mod

```
replace (
	go.etcd.io/etcd/integration/v3 => ./SHOULD_NOT_BE_DEPENDENCY
)
```
A little bit a hack, but very explicit when the error happens.

File: vendor/modules.txt (hunk 7 above)
Comment (ptabor) [reply vendor/modules.txt:30]: Fixing...

File: test (hunk 3 above)
Comment (ptabor) [reply test:53]: Fixing...

**Comment by ptabor** [comment]
> Re: Naming the client module "c". 

I wrote ideas and motivation for different naming options in this section of the doc: 
https://docs.google.com/document/d/19UvKD7by_fEkzLMRi-QNSKEed4kYAW286S3DsNRPDOM/edit#heading=h.rj04ue628kji

Happy to change them - agree that current choice might be sub-optimal.
My intuitions goes toward: go.etcd.io/etcd/client/v3, go.etcd.io/etcd/server/v3, go.etcd.io/etcd/tests/v3

@gyuho @jpbetz @jingyih @cfc4n et al. - please comment (here or in the doc)

**Comment by jpbetz** [comment]
`go.etcd.io/etcd/client/v3`, `go.etcd.io/etcd/server/v3` and `go.etcd.io/etcd/tests/v3` look great to me

**Comment by gyuho** [comment]
> go.etcd.io/etcd/client/v3, go.etcd.io/etcd/server/v3 and go.etcd.io/etcd/tests/v3 look great to me

+1

Thanks for cleaning this up @ptabor

**Comment by philips** [comment]
The motivation doc from @ptabor looks great. The overall outcome seems good.

However, I don’t have time to review. But, it looks like @jpbetz is on the case.

Thanks for working on this!

**Comment by ptabor** [comment]
Thank you. I will let you know when its ready for another pass. 

I'm creating scattered PRs like: 
  * https://github.com/etcd-io/etcd/pull/12279 : getting rid of ./vendor
  * https://github.com/etcd-io/etcd/pull/12286 : making go test --short ./... run only unit tests 
  * [not yet ready] - cleanup of ./test file itself 
That are preparing this refactoring, so having them merged is a dependency.

**Comment by sagikazarmark** [comment]
@ptabor what's the status of this PR? Can I help?

**Comment by ptabor** [comment]
@ptabor sagikazarmark

https://github.com/etcd-io/etcd/pull/12322 (so also 
https://github.com/etcd-io/etcd/pull/12319) are the last 2 PRs, I want to get in before rebaseing this PR,
as they will: 
  
   - make the changes to ./test smaller
   - give me confidence I'm not breaking anything as part of this process.

**Comment by sagikazarmark** [comment]
Cool, thanks for the update!

**Comment by ptabor** [comment]
I'm considering this effort done. 
Getting the release of 3.5.0-alpha.0 (https://github.com/etcd-io/etcd/issues/12498) would be a cherry on top of this.

**Pull request closed without merging**